cubi_tk/irods/check.py

```diff
@@ -84,87 +111,104 @@ def execute(self):
+                return
 
-@retry(wait_fixed=1000, stop_max_attempt_number=3)
-def check_file(file, md5s, req_num_reps, t):
+            if num_files < self.args.num_parallel_tests:
+                s_count = num_files
+            else:
+                s_count = self.args.num_parallel_tests
+            self.irods_sessions += [self.connect() for _ in range(s_count - 1)]
+            pool = ThreadPool(processes=self.args.num_parallel_tests)
+            s_idx = 0
+            for obj in data_objs["files"]:
+                pool.apply_async(
+                    check_file, args=(obj, data_objs["md5s"], self.args.req_num_reps, t)
+                )
+                if s_idx == self.args.num_parallel_tests - 1:
+                    s_idx = 0
+                else:
+                    s_idx += 1
+            pool.close()
+            pool.join()
+
+
+def check_file(data_obj: iRODSDataObject, md5s: dict, req_num_reps: int, t):
     """Perform checks for a single file."""
+    md5_obj = md5s.get(data_obj.path + ".md5")
 
-    # 1) md5 sum file exists?
-    if file + ".md5" not in md5s:
-        e_msg = f"No md5 sum file for: {file}"
+    # 1) MD5 sum file exists?
+    if not md5_obj:
+        e_msg = f"No md5 sum file for: {data_obj.path}"
         logger.error(e_msg)
-        # raise FileNotFoundError(e_msg)
-
-    # 2) enough replicas?
-    try:
-        isysmeta_out = check_output(["isysmeta", "-l", "ls", "{file}"]).decode(sys.stdout.encoding)
-    except SubprocessError as e:  # pragma: nocover
-        logger.error("Problem executing isysmeta: %s (probably retrying)", e)
-        raise
-    meta_info = [
-        {
-            lst[0]: lst[1:]
-            for lst in (entry.split(": ") for entry in repl.splitlines())
-            if len(lst) > 0
-        }
-        for repl in isysmeta_out.split("----")
-    ]
-    if len(meta_info) < req_num_reps:
-        e_msg = f"Not enough ({req_num_reps}) replicates for file: {file}"
-        logger.error(e_msg)
-        # raise FileNotFoundError(e_msg)
-
-    # 3) checksum consistent with .md5 file?
-    try:
-        temp_file = f"./temp_{str(uuid.uuid4())}.md5"
-        check_output(["irsync", "-aK", "i:{file}.md5", temp_file])
-    except SubprocessError as e:  # pragma: nocover
-        logger.error("Could not fetch file for md5 sum check: %s.md5: %s", temp_file, e)
-        raise
-
-    with open(temp_file, "r") as f:
-        md5sum = re.match(r"\S+", f.read()).group(0)
-    os.remove(temp_file)
-
-    if not all(repl["data_checksum"][0] == md5sum for repl in meta_info):
-        logger.error(
-            "File checksum not consistent with md5 file...\n"
-            "file: %s\n.md5-file checksum: %s\n"
-            "metadata checksum: %s",
-            file,
-            md5sum,
-            meta_info[0]["data_checksum"],
+
+    # 2) Checksums of all replicas consistent with .md5 file?
+    else:
+        with md5_obj.open("r") as f:
```

Comment:
Can we use open mode `"rt"` to avoid the decoe below?

Reply:
Maybe? Can try.